### ISSUE TITLE
Einstein Amazeballs

### ISSUE THREAD

**Issue by drphilmarshall** [DarkEnergyScienceCollaboration/Twinkles#50]
@linan7788626, Scott Daniel and @drphilmarshall are going to try and PhoSim a 10 arcsec sky patch containing a massive elliptical galaxy in front of two distant galaxies at different redshifts, to make a double (or perhaps triple) Einstein Ring. At least one of these source galaxies is going to host a supernova, and we'll try and observe it at `enigma_1189` cadence for as long as we have laptop CPU time for. The product will be a black and white YouTube video as well as prototype python and on-the-fly documentation.

**Comment by linan7788626** [comment]
![image](https://cloud.githubusercontent.com/assets/1016652/10856171/e3edcb7a-7f13-11e5-96b9-279f1aaa5aba.png)

**Comment by linan7788626** [comment]
We wanted to do something like this (+PhoSim).

https://www.youtube.com/watch?v=u6mARnNMCUA

**Comment by linan7788626** [comment]
Notes with Phil:

* Rings are too bright, we think: we need to check what units of pixel value `PhoSim` is expecting. I just used a couple of real HST images (whose pixel values are HST counts...)

* The instance catalogs that Scott prepared for us, from `OpSIm`, have dates in the mid 1990's. If we add 10592.25 to each `expmjd`, we get dates that span from 2023/01/01 onwards (exactly 29 years after the `OpSim` period). 

* To annotate the video frames with human-readable dates, we will need an `MJD2date` conversion function.

* There is no lens galaxy in this picture - its as if we have already done perfect lens light subtraction on every image... If we add a plausible lens, it will outshine at least the outer ring and probably the inner one too - so maybe for this fun project its OK to leave out the lens.

* Each image takes ~5 mins to run, perhaps because we are generating too many photons from the unrealistically bright rings. (The sky background is already turned off.)

* There is no sky background in the output `PhoSim` images - we really ought to add some in as a Poisson noise realization. This must be a "standard" `PhoSim` hack, but I don't know where it is written down how to do it.

**Comment by cdfassnacht** [comment]
I believe that there is a standard time package (in astropy, I think) that already has MJD2date-type capabilities.

**Comment by rbiswas4** [comment]
> The instance catalogs that Scott prepared for us, from OpSIm, have dates in the mid 1990's. If we add 10592.25 to each expmjd, we get dates that span from 2023/01/01 onwards (exactly 29 years after the OpSim period).
I think all OpSim output currently has dates from the 1990s. So what you did is a perfectly reasonable thing right now. 

- I am still a little confused about cadence we would like. (a) OpSim cadence or (b) would we like one day cadence. 
- We have the lensed supernovae (position, time, flux)  for each of the lensed cases which was used to create the image. Can we get those together as a file?
- Looking at the youtube video, I see two locations (one on each) in the rings getting having flux varying as a function of time non-monotonically. Can we have a brief description of what is going on here (I don't know enough about the setup or what I should expect).

**Comment by linan7788626** [comment]
Hi @cdfassnacht, 
thank you for your suggestion on MDF2date, I will try astropy first.


Hi @rbiswas4 ,
About the OpSim outputs, @drphilmarshall helped me figuring out how to correct them,
 and @danielsf mailed me his python code to produce these outputs, so I will try to reproduce the new OpSim outputs. Do you mind if we discuss on the annotation after we got new files?

>I am still a little confused about cadence we would like. (a) OpSim cadence or (b) would we like one day >cadence.

I think the answer should be "OpSim cadence". 

>We have the lensed supernovae (position, time, flux) for each of the lensed cases which was used to >create the image. Can we get those together as a file?

I prefer to save them separately, because I would like to producing separate images first.
Of course, we could get them as a file in the future, but so far, keeping them separately is more straightforward to me.

>Looking at the youtube video, I see two locations (one on each) in the rings getting having flux varying >as a function of time non-monotonically. Can we have a brief description of what is going on here (I >don't know enough about the setup or what I should expect).

The youtube video is just a demonstration for time delays in a normal galaxy scale lensing system, not the same as the one we are working on (double Einstein rings). The lensed images in this video are not two Einstein Rings of two different sources, they are multiple lensed images of one source. Orange Ellipsoid stands for the image of foreground lens galaxy. White ball is the image of source, actually we can not observe it in real observations, it is located here for visualizing the correlation between flux variations of the lensed images and the source . Blue arclets are lensed images of the white ball. Since the source galaxy is located around three-images region in the source plane, and the central faint image is contaminated by the light from lens galaxy, we only "observe" two images here. Moreover, I put a "supernova" in the source galaxy, and there are also two lensed "supernovas" (same reason as lensed galaxies). You can see the flux varying first in the source, then in the lensed image at right side because of a lensing time delay.  At last, you can see the flux varying in the lensed image at left side because of an even longer time delay. The time delay between two lensed supernovas is an observable effect which can be used to constrain the cosmological model. As is mentioned above, this video is just a demonstration, the only thing we expect in this video is the difference between the arriving time of lensed "supernovas". In principle, the video we tried to make is similar with this one, but is much more complicated and realistic. Basically, it should be made of a set of figures like the one we made (see the picture above). 
In the latter video, we expect : 
* PSF smearing and noise polluting. 
* Flux variation of Einstein rings because of the status of LSST, atmosphere, and ................
* Flux variation of 4 lensed supernovas because of all the effects above plus lensing time delays.

I hope this would be helpful. 


BTW, I am working on cleaning the codes and files of our hack project, and will push them here asap.

**Comment by rbiswas4** [comment]
@linan7788626 

> I prefer to save them separately, because I would like to producing separate images first.
Of course, we could get them as a file in the future, but so far, keeping them separately is more straightforward to me.

Sure, that is fine, but can you point me to the separately saved files? What I am trying to see is how the different lensed supernovae differ from the original one I gave you.

>  The youtube video is just a demonstration for time delays in a normal galaxy scale lensing system, not the same as the one we are working on (double Einstein rings). ...

I am sure you had mentioned some of this at various times, but being able to read this description while looking at the video is very helpful to me. The question I am asking (I think I can phrase it better with the description in hand) : Is the sequence repeated a few times in the youtube video? It seems like all three of the source and lensed supernovae go from dim, to bright, to dim, to bright, to dim,  .. etc. ie. it looks like multiple lightcurves.

**Comment by linan7788626** [comment]
@rbiswas4 
>Sure, that is fine, but can you point me to the separately saved files? What I am trying to see is how the different lensed supernovae differ from the original one I gave you.

Now, the lensed supernovas are represented using fits files, not catalogs. And I am still working on it. 
I will point you to the images as soon as they are done (hopefully,it would be tomorrow).

>Is the sequence repeated a few times in the youtube video? It seems like all three of the source and lensed supernovae go from dim, to bright, to dim, to bright, to dim, .. etc. ie. it looks like multiple 
light curves.

You are right, in current video, I made the flux of supernova varying periodically, because we do not need to replay it if people want to see it again.  However, in the video we are trying to make, we may show the variation only once. Of course, we can talk whether we should set it periodically or not later.

**Comment by drphilmarshall** [comment]
Hey @linan7788626 - I am going to declare victory on this Hack Day thread, and close this issue out. Can you please issue yourself with getting your Einstein Rings into the Twinkles images? We'll continue the discussion there. Thanks! :-)

**Comment by linan7788626** [comment]
Hi @drphilmarshall - Sure, please feel free to close it. I will open another issue and keep you all updated once I make any progress on involving OpSim and PhoSim into simulated Einstein Rings. 
Thank you :-)

**Comment by drphilmarshall** [comment]
Thanks Nan! I'd issue yourself first, if I were you: it'd good to use the
issues as a group-visible to-do list, so that we can all see what needs to
be done. It's also good to only issue one or two next steps at a time
(because plans often change :-) Thanks!
